### PR TITLE
Improve managing Step counter factory

### PR DESCRIPTION
**Pull request by whylee259**
- Puts `theading.lock` into `IconScoreStepFactory` 
- Reduce setting of the STEP properties(step price, step costs, max step limit) to `IconScoreStepFactory`
- IconScoreStepCounter holds copies of STEP properties in a context

### CODE REVIEW

File: iconservice/icon_service_engine.py [hunk 1]
```diff
@@ -317,17 +330,60 @@ def invoke(self,
...
+                    if event_log.indexed[0] == 'StepCostChanged(str,int)':
+                        precommit_flag |= PrecommitFlag.STEP_COST_CHANGED
+
+                    if event_log.indexed[0] == 'MaxStepLimitChanged(str,int)':
+                        precommit_flag |= PrecommitFlag.STEP_MAX_LIMIT_CHANGED
+
+        return precommit_flag
+
+    def _update_step_properties(self, context, precommit_flag):
+
```
Comment (leeheonseung) [inline iconservice/icon_service_engine.py]: first should call self._push_context(context) before that execute this logic, if you want access properties in SCORE or call call something

Comment (whylee259) [reply iconservice/icon_service_engine.py]: Added it

File: iconservice/icon_service_engine.py [hunk 2]
```diff
@@ -310,15 +321,59 @@ def invoke(self,
...
+            self._push_context(context)
+            governance_score = self._get_governance_score(context)
+
+            step_price = self._get_step_price_from_governance(context, governance_score)
+            context.step_counter.set_step_price(step_price)
+
+            step_costs = self._get_step_costs_from_governance(governance_score)
+            context.step_counter.set_step_costs(step_costs)
+
+            max_step_limits = self._get_step_max_limits_from_governance(governance_score)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
            max_step_limits: dict = self._get_step_max_limits_from_governance(governance_score)
```

File: iconservice/icon_service_engine.py [hunk 3]
```diff
@@ -310,15 +321,59 @@ def invoke(self,
...
+        :param precommit_flag: pre-commit flag
+        """
+        if precommit_flag & PrecommitFlag.STEP_ALL_CHANGED == PrecommitFlag.NONE:
+            return
+
+        try:
+            self._push_context(context)
+            governance_score = self._get_governance_score(context)
+
+            step_price = self._get_step_price_from_governance(context, governance_score)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
            step_price: int = self._get_step_price_from_governance(context, governance_score)
```

File: iconservice/icon_service_engine.py [hunk 4]
```diff
@@ -402,9 +457,7 @@ def _invoke_request(self,
 
         # If the request is V2 the stepLimit field is not there,
         # so fills it as the max step limit to proceed the transaction.
-        step_limit = self._step_counter_factory.get_max_step_limit(context.type)
-        if 'stepLimit' in params:
-            step_limit = min(params['stepLimit'], step_limit)
+        step_limit = params.get('stepLimit', context.step_counter.max_step_limit)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
        step_limit: int = params.get('stepLimit', context.step_counter.max_step_limit)
```

File: iconservice/icon_service_engine.py [hunk 5]
```diff
@@ -505,9 +558,10 @@ def estimate_step(self, request: dict) -> int:
         :return: The amount of step
         """
         context = IconScoreContext(IconScoreContextType.ESTIMATION)
+        context.step_counter = self._step_counter_factory.create(IconScoreContextType.INVOKE)
         # Fills the step_limit as the max step limit to proceed the transaction.
-        step_limit = self._step_counter_factory.get_max_step_limit(IconScoreContextType.INVOKE)
-        context.step_counter = self._step_counter_factory.create(step_limit)
+        step_limit = context.step_counter.max_step_limit
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
        step_limit: int = context.step_counter.max_step_limit
```

File: iconservice/icon_service_engine.py [hunk 6]
```diff
@@ -535,17 +589,17 @@ def query(self, method: str, params: dict) -> Any:
         """
         context = IconScoreContext(IconScoreContextType.QUERY)
         context.block = self._icx_storage.last_block
-        step_limit = self._step_counter_factory.get_max_step_limit(context.type)
+        context.step_counter: IconScoreStepCounter = \
+            self._step_counter_factory.create(IconScoreContextType.QUERY)
+        step_limit = context.step_counter.max_step_limit
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
        step_limit: int = context.step_counter.max_step_limit
```

File: iconservice/icon_service_engine.py [hunk 7]
```diff
@@ -535,17 +589,17 @@ def query(self, method: str, params: dict) -> Any:
...
 
         if params:
             from_: 'Address' = params.get('from', None)
             context.msg = Message(sender=from_)
-            if 'stepLimit' in params:
-                step_limit = min(params['stepLimit'], step_limit)
+            step_limit = params.get('stepLimit', step_limit)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
            step_limit: int = params.get('stepLimit', step_limit)
```

File: iconservice/icon_service_engine.py [hunk 8]
```diff
@@ -569,21 +623,23 @@ def validate_transaction(self, request: dict) -> None:
...
-        step_price: int = self._get_step_price()
-        minimum_step = \
-            self._step_counter_factory.get_step_cost(StepType.DEFAULT)
+
+        context = IconScoreContext(IconScoreContextType.QUERY)
+        context.step_counter: IconScoreStepCounter = \
+            self._step_counter_factory.create(IconScoreContextType.QUERY)
+
+        step_price: int = context.step_counter.step_price
+        minimum_step = self._step_counter_factory.get_step_cost(StepType.DEFAULT)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
        minimum_step: int = self._step_counter_factory.get_step_cost(StepType.DEFAULT)
```

File: iconservice/icon_service_engine.py [hunk 9]
```diff
@@ -569,21 +623,23 @@ def validate_transaction(self, request: dict) -> None:
...
 
         if 'data' in params:
             # minimum_step is the sum of
             # default STEP cost and input STEP costs if data field exists
             data = params['data']
             input_size = self._get_byte_length(data)
-            minimum_step += input_size * \
-                self._step_counter_factory.get_step_cost(StepType.INPUT)
+            minimum_step += input_size * self._step_counter_factory.get_step_cost(StepType.INPUT)
```
Comment (goldworm-icon) [inline iconservice/icon_service_engine.py]:
```suggestion
            minimum_step: int += input_size * self._step_counter_factory.get_step_cost(StepType.INPUT)
```

File: iconservice/iconscore/icon_score_step.py [hunk 10]
```diff
@@ -52,41 +53,62 @@ class IconScoreStepCounterFactory(object):
...
+    def set_step_cost(self, step_type: 'StepType', value: int):
+        """Sets the step cost for specific action.
+
+        :param step_type: specific action
+        :param value: step cost
+        """
+        with self._lock:
+            self._step_costs[step_type] = value
 
     def get_max_step_limit(self, context_type: 'IconScoreContextType'):
```
Comment (goldworm-icon) [inline iconservice/iconscore/icon_score_step.py]:
```suggestion
    def get_max_step_limit(self, context_type: 'IconScoreContextType') -> int:
```

File: iconservice/iconscore/icon_score_step.py [hunk 11]
```diff
@@ -95,20 +117,22 @@ def set_max_step_limit(
...
-        :param step_limit: step limit of the transaction.
+        :param context_type: context type
         :return: step counter
         """
 
-        # Copying a `dict` so as not to change step costs when processing a
-        # transaction.
-        return IconScoreStepCounter(
-            self._step_cost_dict.copy(), step_limit, self._step_price)
+        with self._lock:
```
Comment (goldworm-icon) [inline iconservice/iconscore/icon_score_step.py]: Reduce critical section area with local variables. (step_price, step_costs, max_step_limit)

File: tests/test_icon_score_step.py [hunk 12]
```diff
@@ -523,7 +523,8 @@ def intercept_invoke(*args, **kwargs):
                 # Pass the unknown step type
                 pass
 
-        self.step_counter = IconScoreStepCounter(step_costs, 100, 0)
+        self.step_counter = IconScoreStepCounter(0, step_costs, 100)
+        self.step_counter.reset(100)
```
Comment (goldworm-icon) [inline tests/test_icon_score_step.py]:
```suggestion
        self.step_counter.reset(step_limit=100)
```

File: tests/test_icon_score_step.py [hunk 13]
```diff
@@ -622,7 +623,8 @@ def _calc_step_used(self, offset: int, count: int):
         return step_used
 
     def _assert_step_used(self, step_used: int, request: dict, tx_hash: bytes):
-        self.step_counter = IconScoreStepCounter(self.step_cost_dict, 5000000, 0)
+        self.step_counter = IconScoreStepCounter(0, self.step_cost_dict, 5000000)
```
Comment (goldworm-icon) [inline tests/test_icon_score_step.py]:
```suggestion
        self.step_counter = IconScoreStepCounter(0, self.step_cost_dict, 5_000_000)
```

File: tests/test_icon_score_step.py [hunk 14]
```diff
@@ -622,7 +623,8 @@ def _calc_step_used(self, offset: int, count: int):
...
+        self.step_counter.reset(5000000)
```
Comment (goldworm-icon) [inline tests/test_icon_score_step.py]:
```suggestion
        self.step_counter.reset(step_limit=5_000_000)
```

File: iconservice/icon_service_engine.py (hunk 9 above)
Comment (goldworm-icon) [reply iconservice/icon_service_engine.py]: Ignore this code snippet.

**Pull request merged**